http/apache/nanny.py

```diff
@@ -0,0 +1,318 @@
+    try:
+        with open(instance.pid_file_path) as pid_file:
+            pid = int(pid_file.read())
+    except FileNotFoundError:
+        start_apache_on_port(instance.port)
+        return
+    if not psutil.pid_exists(pid):
+        start_apache_on_port(instance.port)
+
+
+async def main():
+    argument_parser = argparse.ArgumentParser()
+    argument_parser.add_argument(
+        '--listen-port', required=True,
+        help='Port on which to listen for incoming requests'
+    )
+    argument_parser.add_argument(
+        '--server-port-range-start', required=True, type=int,
+        help=(
+            'First port where server runs; further instances run on '
+            'subsequent ports'
+        )
+    )
+    argument_parser.add_argument(
+        '--max-instances', required=False, default=1, type=int,
+        help='Number of server instances to run at a time'
+    )
+    argument_parser.add_argument(
+        '--version', required=True,
+        help='Version identifier of the nanny process. Used to check if already running'
+    )
+    argument_parser.add_argument(
+        '--debug', action='store_true',
+        help='Enable debug logging'
+    )
+    argument_parser.add_argument(
+        '--replace', action='store_true',
+        help='Kill and replace nanny. If not passed, just ensure the nanny is running.'
+    )
+    argument_parser.add_argument(
+        '--stop-running-servers', action='store_true',
+        help=(
+            'Kill all running server instances outside of the port range given'
+        )
+    )
+    argument_parser.add_argument(
+        '--run-dir', required=True, help=(
+            'Path to dir containing pid and version files'
+        ),
+    )
+    args = argument_parser.parse_args()
+    logging.basicConfig()
+    if args.debug:
+        logger.setLevel(logging.DEBUG)
+
+    version = args.version
+    run_dir = pathlib.Path(args.run_dir)
+    pid_path = run_dir / 'nanny.pid'
+    version_path = run_dir / 'version'
+    running_nanny_pid: int
+    if pid_path.is_file():
+        with open(pid_path, 'r') as pid_file:
+            running_nanny_pid = int(pid_file.read().strip())
+    else:
+        running_nanny_pid = -1
+    nanny_running = psutil.pid_exists(running_nanny_pid)
+    if nanny_running and not args.replace and version_path.is_file():
+        with open(version_path, 'r') as version_file:
+            running_nanny_version = version_file.read()
+        if running_nanny_version == version:
+            logger.info(f'Nanny already running with same version')
+            exit(0)
+    with open(version_path, 'w') as version_file:
+        # Save the version before killing the running nanny to prevent
+        # race condition betwen multiples of this script
+        version_file.write(version)
+    if nanny_running:
+        logger.info(f'Killing running nanny process at PID {running_nanny_pid}')
+        os.kill(running_nanny_pid, 9)
+    with open(pid_path, 'w') as pid_file:
```

Comment:
For more idiomatic nanny process, you would need to use some locking system to guard the PID file ([something like this](https://py-filelock.readthedocs.io/en/latest/index.html)). Simply reading the PID file, checking if it's dead, and continuing could cause issues with multiple nanny processes trying to start in parallel. 

Practically, trying to listen on the same socket is equivalent to a lock, as it is a unique resource that would prevent this race condition from causing any actual harm. In a process without that unique resource being used, you'd want to do the above.

If you want to keep using the port as the locking mechanism, I'd use that rather than (in addition to) a PID file. E.g., use the PID file to kill the old nanny if it's a bad version, but first try to claim the port ourselves.

Reply:
@caleblf